### PR TITLE
Improve CM testing for elliptic curves over number fields

### LINKED ISSUE #35147: Improve CM testing for elliptic curves
Currently CM testing for elliptic curves over number fields is implemented in the function `is_cm_j_invariant` with two methods to choose from, both written by me.  These work fine in low degree (say, up to 10) but beyond that become very slow.

In https://arxiv.org/abs/2301.11169 (joint with A.Sutherland) we give a vastly better algorithm, together with implementations in Sage (and gp and Magma), available at https://github.com/AndrewVSutherland/EndECNF.

I will change `is_cm_j_invariant` to use this by default (leaving the other two methods as non-default options).

At the same time I will improve the caching of `cm_j_invariants_and orders` (possibly in a separate PR).

### PR DESCRIPTION
**Pull request by JohnCremona**
### 📚 Description

This implements the Cremona-Sutherland algorithm (see :arxiv:`2301.11169`) for detecting Hilbert Class Polynomials and computing the corresponding CM discriminants.  Using this, a new default algorithm for detecting CM j-invariants, and hence whether or not elliptic curves over number fields, is implemented.  This is very much faster than the older methods for CM testing which were only fast enough to be reasonable over fields of degree less than 10; the new method works fine up to degree 100 and beyond (as some of the examples show).

Secondly, the caching of the computation of CM orders and discriminants and j-invariants for class numbers up to 100 has been made a lot more efficient through better caching.  Before, if you asked for all discriminants with class number (say) 10, it computed all those with class number up to 10 and threw the rest away, so that even if you then asked for class number 9 it had to start from scratch; now that second call will be instantaneous.  And If you then ask for class number 11, it will be faster than it was since much of the data needed will have been cached.

This closes #35147

### 📝 Checklist

- [x ] I have made sure that the title is self-explanatory and the description concisely explains the PR.
- [x ] I have linked an issue or discussion.
- [ x] I have created tests covering the changes.
- [ x] I have updated the documentation accordingly.

### ⌛ Dependencies

None.  Based on 10.0.beta0

### CODE REVIEW

**Comment by JohnCremona** [comment]
I see doctest failures and am fixing them.

I suppose the point of having a "draft" PR is that all the automatic tests will run, then one makes it a real PR after fixing issues.

**Comment by JohnCremona** [comment]
> I see doctest failures and am fixing them.

There are no failures when I run the tests locally so am letting the automatic build and test run again.

**Comment by JohnCremona** [comment]
> > I see doctest failures and am fixing them.
> 
> There are no failures when I run the tests locally so am letting the automatic build and test run again.

OK, there really were, so I have fixed them.

File: src/sage/schemes/elliptic_curves/ell_finite_field.py [hunk 1]
```diff
@@ -1414,6 +1427,150 @@ def _fetch_cached_order(self, other):
...
             self._order = n
 
+    def height_above_floor(self, ell, e):
+        r"""
+        Return the height of the `j`-invariant of this ordinary elliptic curve on its `\ell`-volcano.
+
+        INPUT:
+
+        - ``ell`` -- a prime number
+        - ``e`` -- a non-negative integer
```
Comment (roed314) [inline src/sage/schemes/elliptic_curves/ell_finite_field.py]: The meaning of `e` should be explained: it's not described here.

File: src/sage/schemes/elliptic_curves/ell_finite_field.py [hunk 2]
```diff
@@ -1414,6 +1427,150 @@ def _fetch_cached_order(self, other):
...
+        if self.is_supersingular():
+            raise ValueError("Elliptic curve ({}) must be ordinary".format(self))
+        D1 = self.frobenius_discriminant()
+        D0 = D1.squarefree_part()
+        if D0%4 !=1:
+            D0 *= 4
+        v = ZZ(D1//D0).isqrt()
+        h0 = D0.class_number()
+        if h%h0:
+            return 0
```
Comment (roed314) [inline src/sage/schemes/elliptic_curves/ell_finite_field.py]: I assume these `return 0` statements only happen if the assumptions on the input `h` are not satisfied?  If so, wouldn't a `ValueError` be better?

File: src/sage/schemes/elliptic_curves/ell_finite_field.py [hunk 3]
```diff
@@ -1414,6 +1427,150 @@ def _fetch_cached_order(self, other):
...
+
+        ALGORITHM:
+
+        Compute the trace of Frobenius and hence the discriminant `D_0`
+        and class number `h_0` of the maximal order containing the
+        endomorphism order.  From the given value of `h`, which must be a
+        multiple of `h_0`, compute the possible conductors, using
+        :meth:`height_above_floor` for each prime `\ell` dividing the
+        quotient `h/h_0`.  If exactly one conductor `f` remains, return
+        `f^2D_0`, otherwise return `0`.
```
Comment (roed314) [inline src/sage/schemes/elliptic_curves/ell_finite_field.py]: Will returning 0 ever happen when `h` is correct?

File: src/sage/schemes/elliptic_curves/ell_finite_field.py [hunk 4]
```diff
@@ -1414,6 +1427,150 @@ def _fetch_cached_order(self, other):
...
+        if h%h0:
+            return 0
+        from sage.schemes.elliptic_curves.cm import OrderClassNumber
+        cs = [v//f for f in v.divisors() if OrderClassNumber(D0,h0,f) == h] # cofactors c=v/f compatible with h(f**2D0)=h
+        if not cs:
+            return 0
+        if len(cs) == 1:
+            return (v//cs[0])**2 * D0
+        from sage.sets.set import Set
+        L = sorted(list(Set(sum([c.prime_factors() for c in cs], []))))
```
Comment (roed314) [inline src/sage/schemes/elliptic_curves/ell_finite_field.py]: You can just do
```
L = sorted(set(sum([c.prime_factors() for c in cs], [])))
```

File: src/sage/schemes/elliptic_curves/cm.py [hunk 5]
```diff
@@ -171,6 +174,156 @@ def hilbert_class_polynomial(D, algorithm=None):
...
+    INPUT:
+
+    - ``f`` -- a polynomial in `\ZZ[X]`.
+    - ``check_monic_irreducible`` (boolean, default ``True``) -- if
+      ``True``, check that ``f`` is a monic, irreducible, integer
+      polynomial.
+
+    OUTPUT:
+
+    (integer) -- `D` if ``f`` is the Hilbert Class Polynomial (HCP) `H_D`, else 0.
```
Comment (roed314) [inline src/sage/schemes/elliptic_curves/cm.py]: Missing the boolean part of the output.  Maybe it's best to describe the output in detail here, and just summarize the goal of the method on the first line.

File: src/sage/schemes/elliptic_curves/cm.py [hunk 6]
```diff
@@ -171,6 +174,156 @@ def hilbert_class_polynomial(D, algorithm=None):
...
+    pmin = 33 * (h**2 * (RR(h+2).log().log()+2)**2).ceil()
+    # Guarantees 4*p > |D| for fundamental D under GRH
+    p = pmin-1
+    n = 0
+    from sage.arith.all import next_prime
+    while True:
+        p = next_prime(p)
+        n += 1
+        fp = f.change_ring(GF(p))
+        # Compute X^p-X mod fp manually, avoiding quotient ring which is slower
```
Comment (roed314) [inline src/sage/schemes/elliptic_curves/cm.py]: You can do `pow(z, p, f) - z`, which will be faster (it uses FLINT directly for example).

File: src/sage/schemes/elliptic_curves/cm.py [hunk 7]
```diff
@@ -171,6 +174,156 @@ def hilbert_class_polynomial(D, algorithm=None):
...
+        sage: h = D0.class_number()
+        sage: [OrderClassNumber(D0,h,f) for f in srange(1,20)]
+        [1, 1, 2, 2, 2, 4, 4, 4, 6, 4, 6, 8, 6, 8, 8, 8, 8, 12, 10]
+        sage: all([OrderClassNumber(D0,h,f) == (D0*f**2).class_number() for f in srange(1,20)])
+        True
+
+    """
+    if not D0.is_fundamental_discriminant():
+        raise ValueError("{} is not a fundamental discriminant".format(D0))
+    if not D0.is_fundamental_discriminant() or f <= 0:
```
Comment (roed314) [inline src/sage/schemes/elliptic_curves/cm.py]: You're checking `D0.is_fundamental_discriminant()` twice.

File: src/sage/schemes/elliptic_curves/cm.py [hunk 8]
```diff
@@ -374,232 +536,353 @@ def cm_orders(h, proof=None):
...
+# klaise_table = {}
+# for X in open('klaise_table.txt').readlines():
+#     if len(X.strip()):
+#         v = [int(a) for a in X.split()]
+#         for i in range(4):
+#             klaise_table[v[3*i]]=(v[3*i+2], v[3*i+1])
+
+# The keys are integers 1--100 and the value for each h is (|D|,n)
+# where |D| is the largest discriminant of an imaginary quadratic
+# field with class number h, and n is the number of such fields.
```
Comment (roed314) [inline src/sage/schemes/elliptic_curves/cm.py]: Shouldn't this be "order" rather than "field"?

File: src/sage/schemes/elliptic_curves/cm.py [hunk 9]
```diff
@@ -374,232 +536,353 @@ def cm_orders(h, proof=None):
...
+    by work of Mark Watkins ([Watkins2004]_).
+
+    .. NOTE::
+
+        The class number of a fundamental negative discriminant `D` is
+        the same as the class number of the imaginary quadratic field
+        `\QQ(\sqrt{D})`, so this function gives the number of such
+        fields of each class number `h\le100`.  It is easy to extend
+        this to larger class number conditional on the GRH, but much
+        harder to obyain unconditional results.
```
Comment (roed314) [inline src/sage/schemes/elliptic_curves/cm.py]: typo: obtain.

File: src/sage/schemes/elliptic_curves/cm.py [hunk 10]
```diff
@@ -374,232 +536,353 @@ def cm_orders(h, proof=None):
...
+        sage: largest_disc_with_class_number(10)
+        (13843, 123)
+        sage: largest_disc_with_class_number(100)
+        (1856563, 2311)
+        sage: largest_disc_with_class_number(101)
+        Traceback (most recent call last):
+        ...
+        NotImplementedError: largest discriminant not available for class number 101
+
+    For most `h\le100`, the largest fundamental discriminant with class number `h` is also the largest discriminant, but this is not the case for `h=`::
```
Comment (roed314) [inline src/sage/schemes/elliptic_curves/cm.py]: For which h is this not the case?

Comment (roed314) [reply src/sage/schemes/elliptic_curves/cm.py]: Oh, I see.  Maybe it's better to say `for some h`.

File: src/sage/schemes/elliptic_curves/cm.py [hunk 11]
```diff
@@ -374,232 +536,353 @@ def cm_orders(h, proof=None):
...
+    if h <= 0:
+        return Integer(0), Integer(0)
+    try:
+        B, c = klaise_table[h]
+        return (Integer(B), Integer(c))
+    except KeyError:
+        raise NotImplementedError("largest discriminant not available for class number %s" % h)
+
+# This dict has class numbers h as keys, the value at h is a complete
+# list of paits (D0,f) such that D=D0*f**2 has class number h.  We
```
Comment (roed314) [inline src/sage/schemes/elliptic_curves/cm.py]: typo: pairs

File: src/sage/schemes/elliptic_curves/cm.py [hunk 12]
```diff
@@ -374,232 +536,353 @@ def cm_orders(h, proof=None):
...
     r"""
-    Return dictionary with keys class numbers `h\le hmax` and values the
-    list of all pairs `(D, f)`, with `D<0` a fundamental discriminant such
-    that `Df^2` has class number `h`.  If the optional bound `B` is given,
-    return only those pairs with fundamental `|D| \le B`, though `f` can
-    still be arbitrarily large.
+    Return a dictionary with keys class numbers `h\le hmax` and values the
+    list of all pairs `(D_0, f)`, with `D_0<0` a fundamental discriminant such
+    that `D=D_0f^2` has class number `h`.  If the optional bound `B` is given,
+    return only those pairs with fundamental `|D| \le B`.
```
Comment (roed314) [inline src/sage/schemes/elliptic_curves/cm.py]: you should probably omit "fundamental" here, since you're limiting the discriminant rather than the fundamental discriminant.  Why did you change the behavior on this?  I agree that it makes a bit more sense, but should maybe have a deprecation....

File: src/sage/schemes/elliptic_curves/cm.py [hunk 13]
```diff
@@ -374,232 +536,353 @@ def cm_orders(h, proof=None):
...
         sage: sage.schemes.elliptic_curves.cm.discriminants_with_bounded_class_number(hmax=5, B=50)
-        {1: [(-3, 3), (-3, 2), (-3, 1), (-4, 2), (-4, 1), (-7, 2), (-7, 1), (-8, 1), (-11, 1), (-19, 1), (-43, 1)], 2: [(-3, 7), (-3, 5), (-3, 4), (-4, 5), (-4, 4), (-4, 3), (-7, 4), (-8, 3), (-8, 2), (-11, 3), (-15, 2), (-15, 1), (-20, 1), (-24, 1), (-35, 1), (-40, 1)], 3: [(-3, 9), (-3, 6), (-11, 2), (-19, 2), (-23, 2), (-23, 1), (-31, 2), (-31, 1), (-43, 2)], 4: [(-3, 13), (-3, 11), (-3, 8), (-4, 10), (-4, 8), (-4, 7), (-4, 6), (-7, 8), (-7, 6), (-7, 3), (-8, 6), (-8, 4), (-11, 5), (-15, 4), (-19, 5), (-19, 3), (-20, 3), (-20, 2), (-24, 2), (-35, 3), (-39, 2), (-39, 1), (-40, 2), (-43, 3)], 5: [(-47, 2), (-47, 1)]}
+        {1: [(-3, 1), (-3, 2), (-3, 3), (-4, 1), (-4, 2), (-7, 1), (-7, 2), (-8, 1), (-11, 1), (-19, 1), (-43, 1)], 2: [(-3, 4), (-4, 3), (-8, 2), (-15, 1), (-20, 1), (-24, 1), (-35, 1), (-40, 1)], 3: [(-11, 2), (-23, 1), (-31, 1)], 4: [(-39, 1)], 5: [(-47, 1)]}
+
     """
+    hmax = Integer(hmax)
+    global hDf_dict
+
+    # Easy case where we have already computed and cached the relevant values
+    if hDf_dict and hmax <= max(hDf_dict.keys()):
```
Comment (roed314) [inline src/sage/schemes/elliptic_curves/cm.py]: You can just do `max(hDf_dict)`: the `.keys()` is unnecessary.

File: src/sage/schemes/elliptic_curves/cm.py [hunk 14]
```diff
@@ -374,232 +536,353 @@ def cm_orders(h, proof=None):
...
-
-    # Easy case -- instead of giving error, give meaningful output
-    if hmax < 1:
-        return T
 
     if B is None:
-        # Determine how far we have to go by applying Watkins's theorem.
-        v = [largest_fundamental_disc_with_class_number(h) for h in range(1, hmax+1)]
+        # Determine how far we have to go by applying Watkins + Klaise's results.
+        v = [largest_disc_with_class_number(h) for h in range(1, hmax+1)]
```
Comment (roed314) [inline src/sage/schemes/elliptic_curves/cm.py]: if hmax is larger than 100, it might be better to raise an error telling the user to specify B, rather than a `NotImplementedError`.

File: src/sage/schemes/elliptic_curves/cm.py [hunk 15]
```diff
@@ -374,232 +536,353 @@ def cm_orders(h, proof=None):
...
+    # from the fundamental ones. Note that in the initialisation, the
+    # keys of h_dict include nonfundamental discriminants, but we only
+    # update it with fundamental ones.
+
+    h_dict = {}
+    for h, Dflist in hDf_dict.items():
+        for D0,f in Dflist:
+            h_dict[D0*f**2] = h
+
+    # We do not need to certaify the class number from :pari:`qfbclassno` for discriminants under 2*10^10
```
Comment (roed314) [inline src/sage/schemes/elliptic_curves/cm.py]: typo: certify

File: src/sage/schemes/elliptic_curves/cm.py [hunk 16]
```diff
@@ -374,232 +536,353 @@ def cm_orders(h, proof=None):
...
 
     for h in T:
-        T[h] = list(reversed(T[h]))
+        T[h].sort(key=lambda Df: (Df[0].abs(), Df[1]))
 
-    if fund_count is not None:
-        # Double check that we found the right number of fundamental
-        # discriminants; we might as well, since Watkins provides this
-        # data.
+    # count is None unless the user provided a value of B
```
Comment (roed314) [inline src/sage/schemes/elliptic_curves/cm.py]: count is None precisely when the user provided a value of B.

File: src/sage/schemes/elliptic_curves/cm.py [hunk 17]
```diff
@@ -374,232 +536,353 @@ def cm_orders(h, proof=None):
...
-                        T[h].append(z)
-                    else:
-                        T[h] = [z]
-                f += 1
+                h = OrderClassNumber(D0,h_dict[D0],f)
+
+        # If the class number of this order is within the range, then store (D0,f)
+        if h <= hmax:
+            z = (D0, f)
+            if h in T:
```
Comment (roed314) [inline src/sage/schemes/elliptic_curves/cm.py]: you could use `T = defaultdict(set)` and then this would just be `T[h].add(z)`.

File: src/sage/schemes/elliptic_curves/cm.py [hunk 18]
```diff
@@ -374,232 +536,353 @@ def cm_orders(h, proof=None):
...
+# This dict has class numbers h as keys, the value at h is a complete
+# list of paits (D0,f) such that D=D0*f**2 has class number h.  We
+# initialise it with h=1 only; other values will be added by calls to
+# discriminants_with_bounded_class_number().
 
+hDf_dict = {ZZ(1): [(ZZ(D), ZZ(h)) for D,h in
+                    [(-3, 1), (-3, 2), (-3, 3), (-4, 1), (-4, 2), (-7, 1), (-7, 2),
+                     (-8, 1), (-11, 1), (-19, 1), (-43, 1), (-67, 1), (-163, 1)]]}
 
 @cached_function
```
Comment (roed314) [inline src/sage/schemes/elliptic_curves/cm.py]: You're doing your own caching, so I don't think this should be a cached function anymore.

File: src/sage/schemes/elliptic_curves/cm.py [hunk 19]
```diff
@@ -374,232 +536,353 @@ def cm_orders(h, proof=None):
...
+    # keys of h_dict include nonfundamental discriminants, but we only
+    # update it with fundamental ones.
+
+    h_dict = {}
+    for h, Dflist in hDf_dict.items():
+        for D0,f in Dflist:
+            h_dict[D0*f**2] = h
+
+    # We do not need to certaify the class number from :pari:`qfbclassno` for discriminants under 2*10^10
+    if B > 2*10**10:
```
Comment (roed314) [inline src/sage/schemes/elliptic_curves/cm.py]: Wait, so shouldn't this be the other way?  You can set `proof=False` when B is smaller than `2*10^10`.

File: src/sage/schemes/elliptic_curves/cm.py [hunk 20]
```diff
@@ -374,232 +536,353 @@ def cm_orders(h, proof=None):
...
-    """
-    Return whether or not this is a CM `j`-invariant.
+def is_cm_j_invariant(j, method='CremonaSutherland'):
+    r"""Return whether or not this is a CM `j`-invariant, and the CM discriminant if it is.
 
     INPUT:
 
     - ``j`` -- an element of a number field `K`
 
+    - ``method`` (string, default 'CremonaSutherland') -- the method
```
Comment (roed314) [inline src/sage/schemes/elliptic_curves/cm.py]: You should deprecate the `method` keyword and switch to `algorithm`, which is the standard keyword for this.

File: src/sage/schemes/elliptic_curves/cm.py [hunk 21]
```diff
@@ -641,32 +936,48 @@ def is_cm_j_invariant(j, method='new'):
     if j in QQ:
         return False, None
 
-    # Now j has degree at least 2.  If it is not integral so is not CM:
+    # Next we find its minimal polynomial of j:
+
+    if j.parent().absolute_degree()==2:
```
Comment (roed314) [inline src/sage/schemes/elliptic_curves/cm.py]: should have spaces around `==`.

File: src/sage/schemes/elliptic_curves/cm.py [hunk 22]
```diff
@@ -718,12 +1028,12 @@ def is_cm_j_invariant(j, method='new'):
...
             cmf = fP
         elif cmd != dP: # inconsistent with previous
-            return False, None
+            return (False, None)
         else:           # consistent d, so update f
             cmf = cmf.gcd(fP)
 
-    if cmd==0: # no conclusion, we found no degree 1 primes, revert to old method
-        return is_cm_j_invariant(j, method='old')
+    if cmd==0: # no conclusion, we found no degree 1 primes, revert to default method
```
Comment (roed314) [inline src/sage/schemes/elliptic_curves/cm.py]: Spaces around `==`.

**Review by roed314 — CHANGES_REQUESTED** [review]
Generally looks good!

File: src/sage/schemes/elliptic_curves/ell_finite_field.py (hunk 1 above)
Comment (JohnCremona) [reply src/sage/schemes/elliptic_curves/ell_finite_field.py]: OK

File: src/sage/schemes/elliptic_curves/ell_finite_field.py (hunk 2 above)
Comment (JohnCremona) [reply src/sage/schemes/elliptic_curves/ell_finite_field.py]: Maybe. As we had it the output was undefined if h was wrong (and it says so in the OUTPUT block).  But I'll make the change you suggest,

File: src/sage/schemes/elliptic_curves/ell_finite_field.py (hunk 3 above)
Comment (JohnCremona) [reply src/sage/schemes/elliptic_curves/ell_finite_field.py]: No, and I have changed this to mention raising a ValueError if this happens. And in the place this function is used I now have a try/except instead if testing the return value for being 0.

File: src/sage/schemes/elliptic_curves/ell_finite_field.py (hunk 4 above)
Comment (JohnCremona) [reply src/sage/schemes/elliptic_curves/ell_finite_field.py]: Thanks, I had a feeling that was more complicated than necessary!

File: src/sage/schemes/elliptic_curves/cm.py (hunk 5 above)
Comment (JohnCremona) [reply src/sage/schemes/elliptic_curves/cm.py]: Well spotted.  I decided that reteurning True/False as well as D was redundant, so it just returns D or 0.  I did remember to change the place where this function is used.

File: src/sage/schemes/elliptic_curves/cm.py (hunk 6 above)
Comment (JohnCremona) [reply src/sage/schemes/elliptic_curves/cm.py]: Thanks for this suggestion. I did use to form the quotient ring mod fp and then exponentiate there, but that was slower.

File: src/sage/schemes/elliptic_curves/cm.py (hunk 7 above)
Comment (JohnCremona) [reply src/sage/schemes/elliptic_curves/cm.py]: fixed

File: src/sage/schemes/elliptic_curves/cm.py (hunk 8 above)
Comment (JohnCremona) [reply src/sage/schemes/elliptic_curves/cm.py]: Correct -- cut and paste error -- well spotted

File: src/sage/schemes/elliptic_curves/cm.py (hunk 12 above)
Comment (JohnCremona) [reply src/sage/schemes/elliptic_curves/cm.py]: Right, I deleted "fundamental".   The previous behaviour was unnatural and only implemented that way (by WAS and/or me) because with the previous algorithm that was all you could do. *Please* don't insist on a deprecation!  If you look at the old doctest output it was quite confusing.

This function is not referred to outside this file and could have been hidden (preceding underscore), as it is only called by ``cm_orders`` and ``cm_j_invariants_and_orders``, but I don't want to do that now as it really would require a deprecation.

File: src/sage/schemes/elliptic_curves/cm.py (hunk 14 above)
Comment (JohnCremona) [reply src/sage/schemes/elliptic_curves/cm.py]: Now a ValueError will be raised if `hmax>100` and ``B`` is ``None``.

File: src/sage/schemes/elliptic_curves/cm.py (hunk 19 above)
Comment (JohnCremona) [reply src/sage/schemes/elliptic_curves/cm.py]: Good catch

File: src/sage/schemes/elliptic_curves/cm.py (hunk 20 above)
Comment (JohnCremona) [reply src/sage/schemes/elliptic_curves/cm.py]: Done.

File: src/sage/schemes/elliptic_curves/cm.py (hunk 17 above)
Comment (JohnCremona) [reply src/sage/schemes/elliptic_curves/cm.py]: This is a very neat idea -- and I did not know about defaultdicts, so thanks for that -- but it just required too much additional coding everywhere else, since I rely on the values of hDf_dict being sorted lists, not sets.

Of course I am willing to see a solution based on this idea!

Comment (JohnCremona) [reply src/sage/schemes/elliptic_curves/cm.py]: > This is a very neat idea -- and I did not know about defaultdicts, so thanks for that -- but it just required too much additional coding everywhere else, since I rely on the values of hDf_dict being sorted lists, not sets.
> 
> Of course I am willing to see a solution based on this idea!

OK, the last commit does this (11 insertions(+), 16 deletions(-) :))

**Comment by JohnCremona** [comment]
I have now finished my edits in response to the referee

**Comment by JohnCremona** [comment]
I just fixed a trivial merge conflict which arose because this had been sitting idle for 2 weeks since I addressed all the revier's concerns.  Please @roed314 or someone else update the review so that this can be merged!

**Comment by JohnCremona** [comment]
> This has now been sitting idle for 4 weeks since I addressed all the revier's concerns. Please @roed314 or someone else update the review so that this can be merged!

File: src/sage/schemes/elliptic_curves/cm.py (hunk 12 above)
Comment (roed314) [reply src/sage/schemes/elliptic_curves/cm.py]: Ok

File: src/sage/schemes/elliptic_curves/cm.py (hunk 17 above)
Comment (roed314) [reply src/sage/schemes/elliptic_curves/cm.py]: Great!  I use defaultdicts a lot, so I'm glad the suggestion ended up being useful.

**Comment by roed314** [comment]
Thanks for the reminder!  This looks good to me.

**Review by roed314 — APPROVED** [review]

**Comment by JohnCremona** [comment]
Thanks, @roed314 !
